File: project/sungwoong-onboarding/src/main/java/com/multi/sungwoongonboarding/common/valid/EnumValidator.java
package com.multi.sungwoongonboarding.common.valid;

import jakarta.validation.ConstraintValidator;
import jakarta.validation.ConstraintValidatorContext;


public class EnumValidator implements ConstraintValidator<EnumValid, String> {

    private Class<? extends Enum> enumClass;

    @Override
    public boolean isValid(String value, ConstraintValidatorContext constraintValidatorContext) {
        if (value == null || value.isBlank()) {
            return false;
        }
        try {
            Enum.valueOf(enumClass, value);
            return true;
        } catch (IllegalArgumentException e) {
            return false;
        }
    }

    @Override
    public void initialize(EnumValid constraintAnnotation) {
        this.enumClass = constraintAnnotation.enumClass();
    }
}

File: project/taehyung-onboarding/src/main/kotlin/formService/adapter/port/inbound/web/validator/EnumValidator.kt
package formService.adapter.port.inbound.web.validator

import jakarta.validation.ConstraintValidator
import jakarta.validation.ConstraintValidatorContext

class EnumValidator : ConstraintValidator<IsEnum, String> {
    private lateinit var values: List<String>
    private lateinit var messageTemplate: String

    override fun initialize(constraintAnnotation: IsEnum) {
        values =
            constraintAnnotation.enumClass.java.enumConstants
                .map { it.name }

        messageTemplate = constraintAnnotation.message
    }

    override fun isValid(
        value: String,
        context: ConstraintValidatorContext,
    ): Boolean {
        val isValid = values.contains(value)

        if (!isValid) {
            context.disableDefaultConstraintViolation()
            context
                .buildConstraintViolationWithTemplate(
                    "$messageTemplate [${values.joinToString(", ")}]",
                ).addConstraintViolation()
        }

        return isValid
    }
}
